### PR TITLE
Fix bazel building

### PR DESCRIPTION
**Pull request by coltonmorris**
Changed Bazel to use self managed `node_modules`.

Now other Bazel projects can build the `protoc-gen-ts` plugin using Bazel!

example use case:

**Workspace**
```python
git_repository(
  name = "ts_protoc_gen",
  remote = "https://github.com/improbable-eng/ts-protoc-gen",
  tag = "idk",
)

git_repository(
  name = "io_bazel_rules_go",
  commit = "6bee898391a42971289a7989c0f459ab5a4a84dd", 
  remote = "https://github.com/bazelbuild/rules_go.git",
)
load("@io_bazel_rules_go//go:def.bzl", "go_rules_dependencies", "go_register_toolchains")
go_rules_dependencies()
go_register_toolchains()

http_archive(
  name = "io_bazel_rules_webtesting",
  strip_prefix = "rules_webtesting-master",
  urls = [
    "https://github.com/bazelbuild/rules_webtesting/archive/master.tar.gz",
    ],
)
load("@io_bazel_rules_webtesting//web:repositories.bzl", "browser_repositories", "web_test_repositories")
web_test_repositories()

git_repository(
  name = "build_bazel_rules_nodejs",
  remote = "https://github.com/bazelbuild/rules_nodejs.git",
  tag = "0.9.1",
)
load("@build_bazel_rules_nodejs//:defs.bzl", "node_repositories", "yarn_install")
# Pulled from improbable-eng/ts-protoc-gen repository
node_repositories(package_json = ["@ts_protoc_gen//:package.json"])

git_repository(
    name = "build_bazel_rules_typescript",
    remote = "https://github.com/bazelbuild/rules_typescript.git",
    tag = "0.14.0",
)
load("@build_bazel_rules_typescript//:defs.bzl", "ts_setup_workspace")
ts_setup_workspace()

# installs the node_modules
yarn_install(
  # Note: this must be named 'deps' until a custom rule is written
  name = "deps",
  package_json = "@ts_protoc_gen//:package.json",
  yarn_lock = "@ts_protoc_gen//:yarn.lock",
)
```

Now the `protoc-gen-ts` binary is available using the following Label: `@ts_protoc_gen//bin:protoc-gen-ts`

This PR will ready the repo for a custom Bazel rule that does most of the heavy lifting.

### CODE REVIEW

**Comment by coltonmorris** [comment]
This also indirectly fixes a bug that happens if a Bazel project tries to use this npm package (it has breaking Bazel files and a workspace in it).

File: WORKSPACE [hunk 1]
```diff
@@ -1,21 +1,44 @@
...
+)
+load("@build_bazel_rules_nodejs//:defs.bzl", "node_repositories", "yarn_install")
 node_repositories(package_json = ["//:package.json"])
 
-git_repository(
-    name = "build_bazel_rules_typescript",
-    remote = "https://github.com/bazelbuild/rules_typescript.git",
-    tag = "0.10.0",
+# installs our node_modules
+yarn_install(
```
Comment (jonnyreeves) [inline WORKSPACE]: I'm a little confused by this rule as ts-protoc-gen does not use yarn to manage dependencies? I'm guessing this rule provides interoperability with npm?

File: bin/BUILD [hunk 2]
```diff
@@ -3,10 +3,10 @@ load("@build_bazel_rules_nodejs//:defs.bzl", "nodejs_binary")
 
 nodejs_binary(
     name = "protoc-gen-ts",
+    node_modules = "@deps//:node_modules",
     data = [
-        "@//:node_modules",
-        "@//src",
+      "//src:src",
     ],
-    entry_point = "ts_protoc_gen/src/ts_index",
```
Comment (jonnyreeves) [inline bin/BUILD]: Whoops, good catch. I'm surprised this didn't fail CI, is there a way we can automatically test for errors like this?

File: examples/generated/examplecom/repeated_primitive_message_pb.js [hunk 3]
```diff
@@ -345,14 +345,14 @@ proto.examplecom.RepeatedPrimitiveMessage.prototype.clearMyDoubleList = function
 
 /**
  * repeated float my_float = 2;
- * @return {!Array.<number>}
+ * @return {!Array<number>}
  */
 proto.examplecom.RepeatedPrimitiveMessage.prototype.getMyFloatList = function() {
-  return /** @type {!Array.<number>} */ (jspb.Message.getRepeatedFloatingPointField(this, 2));
+  return /** @type {!Array<number>} */ (jspb.Message.getRepeatedFloatingPointField(this, 2));
```
Comment (jonnyreeves) [inline examples/generated/examplecom/repeated_primitive_message_pb.js]: Which version of protoc are you using? I ask because ideally this generated code should not have changed as it is unrelated to your bazel changes.

Our CI currently uses protoc v3.2.0.

File: package.json [hunk 4]
```diff
@@ -35,7 +35,6 @@
     "@types/lodash": "^4.14.106",
     "@types/mocha": "^2.2.46",
     "@types/node": "^7.0.52",
-    "@types/text-encoding": "0.0.32",
```
Comment (jonnyreeves) [inline package.json]: 👍

File: generate.sh [hunk 5]
```diff
@@ -6,10 +6,12 @@ set -e
 EXAMPLES_GENERATED_DIR=examples/generated
 
 echo "Ensuring we have NPM packages installed..."
-npm install
+# npm install
+yarn
```
Comment (jonnyreeves) [inline generate.sh]: Why the change from npm to yarn? 

I have a strong (mostly personal) for this project to remain using npm.

File: bin/BUILD (hunk 2 above)
Comment (coltonmorris) [reply bin/BUILD]: I'm actually surprised it wasn't caught either, not sure why, and I don't really have the resources to tell you why.
The previous change `bazel build //...:all` is a good way to ensure bazel builds - it would also end up running tests you eventually create with bazel.

File: generate.sh (hunk 5 above)
Comment (coltonmorris) [reply generate.sh]: As mentioned earlier, rules_nodejs recommends it.

File: WORKSPACE (hunk 1 above)
Comment (coltonmorris) [reply WORKSPACE]: Just pushed the commit to use rules_nodejs' npm_install. Works perfectly! 😝

**Comment by easyCZ** [comment]
This is awesome, thanks for contributing!
